### ISSUE TITLE
dependency('foo', static: true) not working as expected

### ISSUE THREAD

**Issue by bigon** [mesonbuild/meson#1709]
Hi,

When using `dependency('foo', static: true)` with pkg-config backend, meson is not trying to statically link the executable against that library

**Comment by nirbheek** [comment]
Yes, it seems we don't pass the correct linker flags for forcing a static link with a particular dependency. It's not trivial to do this either because the only way to make `ld` statically link with an `-lfoo` library is to actually pass the full path to the `foo.a` file.

**Comment by bigon** [comment]
According to https://stackoverflow.com/questions/6578484/telling-gcc-directly-to-link-a-library-statically, you can also use `-l:libfoo.a` might be easier than passing the full path?

**Comment by nirbheek** [comment]
Interesting, that option is not documented anywhere. Any idea which versions of `ld` support it? Does it work on macOS?

**Comment by nirbheek** [comment]
It looks like it is documented, but doesn't say which version added it:

> -l namespec
> --library=namespec
> Add the archive or object file specified by namespec to the list of files to link.  This option may be used any number of times.  If namespec is of the form `:filename`, ld will search the library path for a file called filename, otherwise it will search the library path for a file called `libnamespec.a`.

**Comment by germandiagogomez** [comment]
I am trying to link SDL and boost statically and I am having exactly the same problem, which causes some headaches for deployment. Is this supported in version 0.40.0? Thanks for your time.

**Comment by barcharcraz** [comment]
Should this not pass -Bstatic before the libs?

**Comment by barcharcraz** [comment]
indeed I modified dependencies/base.py to add -Wl,-Bstatic to the command line before adding libs if static is on. and -Wl,-Bdynamic afterwards. Gotta make sure it works with msvc though

**Comment by jpakkane** [comment]
The correct fix would be to pass `--static` to pkg-config and use the output as is. If pkg-config gives incorrect arguments, it is a bug in pkg-config or the dependency's pc files and should be fixed there.

**Comment by barcharcraz** [comment]
Meson already does that. However static does not include these flags
seemingly as policy. It’s certantly an issue that can be brought to the
pkg-config devs.

It seems like the —static flag to pkg-config is for linking with the
—static flag to gcc. It’s possible that trying to use the —static gcc flag
and the ending -Bdynamic won’t work.

**Comment by Volker-Weissmann** [comment]
This bug was fixed by 9da031c04c9905949c896a8fdc5b608acd6dde47

Somebody please close this issue.


Test on current master branch:
```meson
project('hello', 'c')
zlib_dep = dependency('zlib', static : true, method: 'pkg-config')
exe = executable('hello', 'hello.c', dependencies: zlib_dep)
```
```
[1/2] cc -Ihello.p -I. -I.. -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -O0 -g -MD -MQ hello.p/hello.c.o -MF hello.p/hello.c.o.d -o hello.p/hello.c.o -c ../hello.c
[2/2] cc  -o hello hello.p/hello.c.o -Wl,--as-needed -Wl,--no-undefined -Wl,--start-group /usr/lib/libz.a -Wl,--end-group
```
